rust/src/countme/cookie.rs

```diff
@@ -0,0 +1,183 @@
+pub fn new() -> Cookie {
+    let now = Utc::now().timestamp();
+    // Start default window at COUNTME_OFFSET to avoid negative values
+    let c = Cookie {
+        epoch: now,
+        window: COUNTME_OFFSET,
+        now,
+    };
+    let cookie: path::PathBuf = [STATE_DIR, COUNTME_COOKIE].iter().collect();
+    if !cookie.exists() {
+        return c;
+    }
+    let file = match fs::File::open(&cookie) {
+        Err(e) => {
+            println!("Could not open '{}': {}", cookie.display(), e);
+            return c;
+        },
+        Ok(f) => f,
+    };
+    let mut reader = BufReader::new(file);
+    let mut line = String::new();
+    // We only read the first line and ignore everything else from the file
+    match reader.read_line(&mut line) {
+        Err(e) => {
+            println!("Could not read from '{}': {}", cookie.display(), e);
+            return c;
+        },
+        Ok(_c) => (),
+    };
+    let version = match line.split(' ').next() {
+        None => {
+            println!("Could not read cookie version from '{}'", cookie.display());
+            return c;
+        },
+        Some(s) => s,
+    };
+    match version {
+        "0" => {
+            match parse_cookie_v0(&line, now) {
+                None => {
+                    println!("Could not parse as version 0 from '{}'", cookie.display());
+                    c
+                }
+                Some(c) => c,
+            }
+        },
+        _ => {
+            println!("Unknown version '{}'", cookie.display());
+            c
+        },
+    }
+}
+
+impl Cookie {
+    fn epoch_window(&self) -> i64 {
+        (self.epoch - COUNTME_OFFSET) / COUNTME_WINDOW
+    }
+
+    fn previous_window(&self) -> i64 {
+        (self.window - COUNTME_OFFSET) / COUNTME_WINDOW
+    }
+
+    fn current_window(&self) -> i64 {
+        (self.now - COUNTME_OFFSET) / COUNTME_WINDOW
+    }
+
+    /// Are we still in the same window as the one we loaded from the cookie?
+    pub fn existing_window(&self) -> bool {
+        self.current_window() <= self.previous_window()
+    }
+
+    // Count Me window logic
+    // https://dnf.readthedocs.io/en/latest/conf_ref.html?highlight=countme#options-for-both-main-and-repo
+    pub fn get_window_counter(&self) -> i64 {
+        // Start counting windows from 1
+        let mut counter = self.current_window() - self.epoch_window() + 1;
+        if counter <= 0 {
+            counter = 1;
+        }
+        match counter {
+            1 => 1,
+            2..=4 => 2,
+            5..=24 => 3,
+            _ => 4, // >= 25
+        }
+    }
+
+    /// Update cookie timestamp
+    pub fn persist(&self) -> Result<()> {
+        let cookie: path::PathBuf = [STATE_DIR, COUNTME_COOKIE].iter().collect();
+        let mut file = fs::File::create(&cookie)?;
+        let content = format!("{} {} {}", COUNTME_VERSION, self.epoch, self.now);
+        file.write_all(content.as_bytes())
```

Comment:
This is all just:
```
let path = Path::new(STATE_DIR).join(COUNTME_COOKIE);
fs::write(&path, format!("{} {} {}", COUNTME_VERSION, self.epoch, self.now))?;
Ok(())
```

(Notice we're also checking for errors on write, you weren't checking the result of `write_all()`)